packages/cojson-storage/src/managerAsync.ts

```diff
@@ -42,113 +47,98 @@ export class StorageManagerAsync {
         await this.handleContent(msg);
         break;
       case "known":
-        await this.handleKnown(msg);
+        this.handleKnown(msg);
         break;
       case "done":
-        await this.handleDone(msg);
+        this.handleDone(msg);
         break;
     }
   }
 
   async sendNewContent(
     coValueKnownState: CojsonInternalTypes.CoValueKnownState,
-  ): Promise<void> {
-    const outputMessages: OutputMessageMap =
-      await this.collectCoValueData(coValueKnownState);
-
-    // reverse it to send the top level id the last in the order
-    const collectedMessages = Object.values(outputMessages).reverse();
-    for (const { knownMessage, contentMessages } of collectedMessages) {
-      this.sendStateMessage(knownMessage);
-
-      if (contentMessages?.length) {
-        for (const msg of contentMessages) {
-          this.sendStateMessage(msg);
-        }
-      }
-    }
-  }
-
-  private async collectCoValueData(
-    peerKnownState: CojsonInternalTypes.CoValueKnownState,
-    messageMap: OutputMessageMap = {},
-    asDependencyOf?: CojsonInternalTypes.RawCoID,
   ) {
-    if (messageMap[peerKnownState.id]) {
-      return messageMap;
-    }
-
-    const coValueRow = await this.dbClient.getCoValue(peerKnownState.id);
+    const coValueRow = await this.dbClient.getCoValue(coValueKnownState.id);
 
     if (!coValueRow) {
       const emptyKnownMessage: KnownStateMessage = {
         action: "known",
-        ...emptyKnownState(peerKnownState.id),
+        ...emptyKnownState(coValueKnownState.id),
       };
-      if (asDependencyOf) {
-        emptyKnownMessage.asDependencyOf = asDependencyOf;
-      }
-      messageMap[peerKnownState.id] = { knownMessage: emptyKnownMessage };
-      return messageMap;
+
+      this.sendStateMessage(emptyKnownMessage);
+      return;
     }
 
     const allCoValueSessions = await this.dbClient.getCoValueSessions(
       coValueRow.rowID,
     );
 
-    const newCoValueKnownState: CojsonInternalTypes.CoValueKnownState = {
-      id: coValueRow.id,
-      header: true,
-      sessions: {},
-    };
+    const signaturesBySession = new Map<
+      SessionID,
+      Pick<SignatureAfterRow, "idx" | "signature">[]
+    >();
 
+    let contentStreaming = false;
     for (const sessionRow of allCoValueSessions) {
-      newCoValueKnownState.sessions[sessionRow.sessionID] = sessionRow.lastIdx;
+      const signatures = await this.dbClient.getSignatures(sessionRow.rowID, 0);
+
+      if (signatures.length > 0) {
+        contentStreaming = true;
+        signaturesBySession.set(sessionRow.sessionID, signatures);
+      }
     }
 
-    const newContentMessages: CojsonInternalTypes.NewContentMessage[] = [
-      {
-        action: "content",
+    /**
+     * If we are going to send the content in streaming, we send before a known state message
+     * to let the peer know how many transactions we are going to send.
+     */
+    if (contentStreaming) {
```

Comment:
We really need to send the known state only if we are going to do streaming, otherwise it's useless because the content we will send already contains our known state.